cmd/juju/application/bundle.go

```diff
@@ -1141,3 +1213,137 @@ func processBundleConfig(data *charm.BundleData, bundleConfigFile string) error
+		}
+		for unitName, unit := range appStatus.Units {
+			application.Units = append(application.Units, bundlechanges.Unit{
+				Name:    unitName,
+				Machine: unit.Machine,
+			})
+		}
+		applications[name] = application
+		annotationTags = append(annotationTags, names.NewApplicationTag(name).String())
+		appNames = append(appNames, name)
+	}
+	model := &bundlechanges.Model{
+		Applications: applications,
+		Machines:     machines,
+	}
+	for _, relation := range status.Relations {
+		// All relations have two endpoints except peers.
+		if count := len(relation.Endpoints); count != 2 {
+			continue
+		}
+		model.Relations = append(model.Relations, bundlechanges.Relation{
+			App1:      relation.Endpoints[0].ApplicationName,
+			Endpoint1: relation.Endpoints[0].Name,
+			App2:      relation.Endpoints[1].ApplicationName,
+			Endpoint2: relation.Endpoints[1].Name,
+		})
+	}
+	// Get all the annotations.
+	annotations, err := apiRoot.GetAnnotations(annotationTags)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, result := range annotations {
+		if result.Error.Error != nil {
+			return nil, errors.Trace(result.Error.Error)
+		}
+		tag, err := names.ParseTag(result.EntityTag)
+		if err != nil {
+			return nil, errors.Trace(err) // This should never happen.
+		}
+		switch kind := tag.Kind(); kind {
+		case names.ApplicationTagKind:
+			model.Applications[tag.Id()].Annotations = result.Annotations
+		case names.MachineTagKind:
+			model.Machines[tag.Id()].Annotations = result.Annotations
+		default:
+			return nil, errors.Errorf("unexpected tag kind for annotations: %q", kind)
+		}
+	}
+	// Now get all the application config.
+	configValues, err := apiRoot.GetConfig(appNames...)
+	if err != nil {
+		return nil, errors.Annotate(err, "getting application options")
+	}
+	for i, config := range configValues {
+		options := make(map[string]interface{})
+		// The config map has values that looks like this:
+		//  map[string]interface {}{
+		//        "value":       "",
+		//        "source":     "user", // or "unset" or "default"
+		//        "description": "Where to gather metrics from.\nExamples:\n  host1.maas:9090\n  host1.maas:9090, host2.maas:9090\n",
+		//        "type":        "string",
+		//    },
+		// We want the value iff default is false.
+		for key, valueMap := range config {
+			value, err := applicationConfigValue(key, valueMap)
+			if err != nil {
+				return nil, errors.Annotatef(err, "bad application config for %q", appNames[i])
+			}
+			if value != nil {
+				options[key] = value
+			}
+		}
+		model.Applications[appNames[i]].Options = options
+	}
+	// Lastly get all the application constraints.
+	constraintValues, err := apiRoot.GetConstraints(appNames...)
+	if err != nil {
+		return nil, errors.Annotate(err, "getting application constraints")
+	}
+	for i, value := range constraintValues {
+		model.Applications[appNames[i]].Constraints = value.String()
+	}
+
+	model.ConstraintsEqual = func(a, b string) bool {
```

Comment:
Because the model type is in the bundlechanges library, and constraints are a juju thing that is in the juju tree. Pushing the understanding of constraints into the bundlechanges library ties them too closely together.